### ISSUE TITLE
Revisit copyable types without equality

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp#2310]
[Regular Types](http://stepanovpapers.com/DeSt98.pdf) in C++ are easier to use and provide a good model to strive for in type design. One of the implications of regular types is that assignment implies some notion of equality. That is, assignment has a post condition that the types are equal. In code this is:

```cc
T a = b;
asert(a == b);  // Post condition of assignment
```

The contrapositive of this is that if equality cannot be defined for a type, then perhaps assignment doesn't make sense either.

I'm not suggesting that we must *define* equality for all of our assignable types. But if we have types where equality might not logically make sense, that might also imply that assignment doesn't logically make sense either. And if that's the case, we may want to remove assignment (though move is fine).

For example, `bigtable::TableAdmin` (https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/bigtable/table_admin.h) explicitly enables assignment, but not equality. That's not necessarily a problem itself (perhaps we just didn't get around to adding equality?). But if equality is not defined on TableAdmin because it logically does not make sense, then one could argue that assignment doesn't make sense either and should likely be `=delete`'d. In this particular case, I do not know what equality of TableAdmin instances would mean, so I suspect that in this particular case we may want to remove copy assignment.

There may be other classes where we want to revisit this too. We should also keep this principle in mind when designing new classes.

## TL;DR:

* Copy-assignment implies there should be a notion of equality.
* Move-only assignment does not necessarily imply a notion of equality.
* If a type has no sensible definition for equality, then it probably shouldn't support copy-assignment either.

**Comment by devjgm** [comment]
How can we find types that define copy-assignment without equality? Here's a rough approximation by finding files that define a defaulted assignment operator and there's no `op==` defined in the same file.

```console
$ ack '\boperator=\(.+default' --cc | grep -v "&&" | grep -v bigquery | cut -f1 -d: | sort | uniq | xargs grep -L "operator=="
google/cloud/bigtable/column_family.h
google/cloud/bigtable/filters.h
google/cloud/bigtable/mutations.h
google/cloud/bigtable/read_modify_write_rule.h
google/cloud/bigtable/row_set.h
google/cloud/bigtable/table_admin.h
google/cloud/internal/filesystem.h
google/cloud/pubsub/publisher.h
```

**Comment by devjgm** [comment]
After #7724, the only obviously remaining copyable files that don't support equality are in bigtable:
```
google/cloud/bigtable/column_family.h
google/cloud/bigtable/filters.h
google/cloud/bigtable/mutations.h
google/cloud/bigtable/read_modify_write_rule.h
google/cloud/bigtable/row_set.h
google/cloud/bigtable/table_admin.h
```
Adding op== shouldn't be very difficult (see  #7724 as an example). This may be something that @dbolduc wants to consider for the Bigtable work, and perhaps as part of the modernization work.

**Comment by devjgm** [comment]
While I maintain that it's good practice to define equality for types that have assignment, it seems incredibly low priority. I'm marking this as a good first issue. And the next time we review this issue, if it's not done, we might as well just close it.

**Comment by dbolduc** [comment]
Modifying the search reveals some other cases. `google::cloud::Options` is an interesting one.
```
$ ack '\boperator=\(' --cc | grep -v "&&" | grep -v delete | cut -f1 -d: | sort | uniq | xargs grep -L "operator=="
bigtable/column_family.h
bigtable/filters.h
bigtable/mutations.h
bigtable/read_modify_write_rule.h
bigtable/row_set.h
bigtable/table_admin.h
options.h
pubsublite/internal/resumable_async_streaming_read_write_rpc.h
spanner/create_instance_request_builder.h
spanner/update_instance_request_builder.h
storage/iam_policy.h
storage/internal/patch_builder.h
```

**Comment by dbolduc** [comment]
The only remaining type that defines an assignment operator, but no equality operator is `google::cloud::Options`. Even if the task is possible, it is too difficult for me. Closing.
